File: sdk/csharp/RaptoreumSdk/RaptoreumClient.cs
using System;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace RaptoreumSdk
{
    public class RaptoreumRPCException : Exception
    {
        public int Code { get; }
        public RaptoreumRPCException(int code, string message) : base($"RPC Error [{code}]: {message}")
        {
            Code = code;
        }
    }

    public class RaptoreumClient
    {
        private readonly HttpClient _httpClient;
        private readonly string _baseUrl;

        public RaptoreumClient(string host = "127.0.0.1", int port = 8766, string user = "", string password = "", bool useSsl = false)
        {
            var scheme = useSsl ? "https" : "http";
            _baseUrl = $"{scheme}://{host}:{port}/";

            _httpClient = new HttpClient();
            _httpClient.DefaultRequestHeaders.Accept.Clear();
            _httpClient.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

            if (!string.IsNullOrEmpty(user) || !string.IsNullOrEmpty(password))
            {
                var authBytes = Encoding.ASCII.GetBytes($"{user}:{password}");
                _httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Basic", Convert.ToBase64String(authBytes));
            }
        }

        public async Task<JsonElement> RequestAsync(string method, params object[] parameters)
        {
            var payload = new
            {
                jsonrpc = "1.0",
                id = "rtm-sdk-csharp",
                method = method,
                @params = parameters ?? new object[] { }
            };

            var jsonPayload = JsonSerializer.Serialize(payload);
            var content = new StringContent(jsonPayload, Encoding.UTF8, "application/json");

            var response = await _httpClient.PostAsync(_baseUrl, content);
            if (!response.IsSuccessStatusCode && response.StatusCode != System.Net.HttpStatusCode.InternalServerError)
            {
                throw new HttpRequestException($"HTTP error status code: {response.StatusCode}");
            }

            var responseString = await response.Content.ReadAsStringAsync();
            using var doc = JsonDocument.Parse(responseString);
            var root = doc.RootElement.Clone();

            if (root.TryGetProperty("error", out var errorProp) && errorProp.ValueKind != JsonValueKind.Null)
            {
                var code = errorProp.GetProperty("code").GetInt32();
                var message = errorProp.GetProperty("message").GetString() ?? "Unknown RPC Error";
                throw new RaptoreumRPCException(code, message);
            }

            return root.GetProperty("result").Clone();
        }

        // Blockchain API
        public async Task<JsonElement> GetBlockchainInfoAsync() => await RequestAsync("getblockchaininfo");
        public async Task<JsonElement> GetBlockCountAsync() => await RequestAsync("getblockcount");
        public async Task<JsonElement> GetBlockHashAsync(int height) => await RequestAsync("getblockhash", height);
        public async Task<JsonElement> GetBlockAsync(string hash, int verbosity = 1) => await RequestAsync("getblock", hash, verbosity);

        // Wallet API
        public async Task<JsonElement> GetBalanceAsync() => await RequestAsync("getbalance");
        public async Task<JsonElement> GetNewAddressAsync(string label = "", string addressType = "legacy") => await RequestAsync("getnewaddress", label, addressType);
        public async Task<JsonElement> ValidateAddressAsync(string address) => await RequestAsync("validateaddress", address);
        public async Task<JsonElement> SendManyAsync(System.Collections.Generic.Dictionary<string, double> amounts, int minconf = 1, string comment = "") => await RequestAsync("sendmany", "", amounts, minconf, comment);
    }
}

File: sdk/fsharp/RaptoreumSdk/RaptoreumClient.fs
namespace RaptoreumSdk

open System
open System.Net.Http
open System.Net.Http.Headers
open System.Text
open System.Text.Json
open System.Threading.Tasks

type RaptoreumRPCException(code: int, message: string) =
    inherit Exception(sprintf "RPC Error [%d]: %s" code message)
    member this.Code = code

type RaptoreumClient(host: string, port: int, user: string, password: string, useSsl: bool) =
    let scheme = if useSsl then "https" else "http"
    let baseUrl = sprintf "%s://%s:%d/" scheme host port
    let httpClient = new HttpClient()

    do
        httpClient.DefaultRequestHeaders.Accept.Clear()
        httpClient.DefaultRequestHeaders.Accept.Add(MediaTypeWithQualityHeaderValue("application/json"))
        if not (String.IsNullOrEmpty(user)) || not (String.IsNullOrEmpty(password)) then
            let authBytes = Encoding.ASCII.GetBytes(sprintf "%s:%s" user password)
            httpClient.DefaultRequestHeaders.Authorization <- 
                AuthenticationHeaderValue("Basic", Convert.ToBase64String(authBytes))

    new() = RaptoreumClient("127.0.0.1", 8766, "", "", false)

    member this.RequestAsync(methodName: string, [<ParamArray>] parameters: obj[]) : Task<JsonElement> =
        task {
            let payload = {| jsonrpc = "1.0"; id = "rtm-sdk-fsharp"; method = methodName; ``params`` = parameters |}
            let jsonPayload = JsonSerializer.Serialize(payload)
            use content = new StringContent(jsonPayload, Encoding.UTF8, "application/json")
            
            let! response = httpClient.PostAsync(baseUrl, content)
            if not response.IsSuccessStatusCode && response.StatusCode <> System.Net.HttpStatusCode.InternalServerError then
                raise (new HttpRequestException(sprintf "HTTP error status code: %O" response.StatusCode))
                
            let! responseString = response.Content.ReadAsStringAsync()
            use doc = JsonDocument.Parse(responseString)
            let root = doc.RootElement.Clone()
            
            let mutable errProp = new JsonElement()
            if root.TryGetProperty("error", &errProp) && errProp.ValueKind <> JsonValueKind.Null then
                let code = errProp.GetProperty("code").GetInt32()
                let message = errProp.GetProperty("message").GetString()
                raise (new RaptoreumRPCException(code, message))
                
            return root.GetProperty("result").Clone()
        }

    member this.GetBlockchainInfoAsync() = this.RequestAsync("getblockchaininfo", [||])
    member this.GetBlockCountAsync() = this.RequestAsync("getblockcount", [||])
    member this.GetBalanceAsync() = this.RequestAsync("getbalance", [||])
    member this.ValidateAddressAsync(address: string) = this.RequestAsync("validateaddress", [| address |])
    member this.SendManyAsync(amounts: System.Collections.Generic.IDictionary<string, double>, minconf: int, comment: string) = 
        this.RequestAsync("sendmany", [| "" :> obj; amounts :> obj; minconf :> obj; comment :> obj |])
